File: ruby/lib/forty_time.rb
# frozen_string_literal: true

class FortyTime
  VERSION = '0.0.5'

  class ParseError < StandardError; end

  def self.parse(input)
    return NullFortyTime.new if input.nil?
    return new(input) if input.is_a? Integer

    can_parse_string = input.is_a?(String) && input.match?(/:/)
    raise ParseError unless can_parse_string

    minutes = parse_string(input)
    new(minutes)
  end

  def self.parse_string(string)
    hours, extra = string.split(':').map(&:to_i)
    extra *= -1 if string[0] == '-'

    hours * 60 + extra
  end

  attr_accessor :minutes

  def initialize(minutes)
    @minutes = minutes
  end

  def +(other)
    result = @minutes + other.minutes
    FortyTime.new(result)
  end

  def -(other)
    result = @minutes - other.minutes
    FortyTime.new(result)
  end

  def to_s
    hours = (@minutes / 60.0).to_i
    extra = (@minutes - hours * 60.0).to_i

    if @minutes.negative?
      hours *= -1
      extra *= -1
      hours = "-#{hours}"
    end

    extra = "0#{extra}" if extra < 10
    [hours, extra].join(':')
  end
end

require_relative './null_forty_time'

File: typescript/src/FortyTime.ts
import { BaseFortyTime } from "./BaseFortyTime"
import { NullFortyTime } from "./NullFortyTime"

export class FortyTime extends BaseFortyTime {
  static parse = (input: number | string | null) => {
    if (input === null) {
      return new NullFortyTime()
    }

    if (typeof input === "number") {
      if (Number.isInteger(input)) {
        return new FortyTime(input)
      } else {
        throw new FortyTime.ParseError()
      }
    }

    if (!input.match(/:/)) throw new FortyTime.ParseError()

    let [hours, extra] = input.split(":").map(Number)

    if (input[0] === "-") {
      extra = extra * -1
    }

    let minutes = hours * 60 + extra

    return new FortyTime(minutes)
  }

  toString = (): string => {
    let hours: string | number = Math.trunc(this.minutes / 60)
    let extra: string | number = this.minutes - hours * 60

    if (this.minutes < 0) {
      hours = hours * -1
      extra = extra * -1
      hours = `-${hours}`
    }

    if (extra < 10) {
      extra = `0${extra}`
    }

    return [hours, extra].join(":")
  }

  plus = (other: BaseFortyTime): BaseFortyTime => {
    const sum = this.minutes + other.minutes
    const result = new FortyTime(sum)
    return result
  }

  minus = (other: BaseFortyTime): BaseFortyTime => {
    const diff = this.minutes - other.minutes
    const result = new FortyTime(diff)
    return result
  }
}
